File: os023l1/TwoLevelPaging.java
import java.util.Scanner;

class TwoLevelPaging {

	long inputAddress(String msg){

		Scanner sc=new Scanner(System.in);
		System.out.print(msg);
		long address = sc.nextLong();

		return address;
	}


	String convert2Binary(long number){
		String b="";
		while (number > 0){
			b = number % 2 + b;
			number /= 2;
		}
		return b;
	}

	int convert2Decimal(String s){
		int d = 0;
		int l = s.length();
		int y=0;
		for (int i=l;i>0;i--){
			d += Long.parseLong(s.substring(i-1,i))*power(2,y++);
		}
		return d;
	}

	long power(int x,int y){

		long p=1;
		for (int i=y;i>0;i--){
			p *= x;
		}

		return p;

	}

	String padLeft(String s){

		int l=s.length();
		for (int i=1;i<=32-l;i++){
			s="0"+s;
		}

		return s;

	}

	public static void main(String[] args){

		TwoLevelPaging tlp=new TwoLevelPaging();
		long address=tlp.inputAddress("Please input logical address: ");
		System.out.println("Address : "+address);
		String binary = tlp.convert2Binary(address);
		binary = tlp.padLeft(binary);
		System.out.println(binary);
		String p1=binary.substring(0,10);
		String p2=binary.substring(10,20);
		String d=binary.substring(20,32);

		System.out.println("p1 : "+p1);
		System.out.println("p2 : "+p2);
		System.out.println(" d : "+d);
		System.out.println("p1 : "+tlp.convert2Decimal(p1));
		System.out.println("p2 : "+tlp.convert2Decimal(p2));
		System.out.println(" d : "+tlp.convert2Decimal(d));


	}
}

File: os023l1/python/two_level_paging.py
__author__ = 'ariel'


def convert_to_binary(number):

    print ("number: ",number)
    b = ""
    while number > 0:
        b = str(number % 2) + b
        number = int(number / 2)

    print ("binary ", b)

    return b


def convert_to_decimal(s):

    d = 0
    l = len(s)
    y = 0
    i = l
    while i > 0:

        d += int(s[i-1:1]) * 2**y
        y += 1

    return d


def pad_left(s):

    return (32-len(s))*'0'+s


def main():

    address = long(raw_input("Please input logical address: "))

    print("Address : ", address)

    binary = convert_to_binary(address)
    binary = pad_left(binary)

    print(binary)
    p1 = binary[0:12]
    p2 = binary[12:22]
    d = binary[22:32]

    print("p1 : ", p1)
    print("p2 : ", p2)
    print(" d : ", d)
    print("p1 : ", convert_to_decimal(p1))
    print("p2 : ", convert_to_decimal(p2))
    print(" d : ", convert_to_decimal(d))



if __name__ == '__main__':
    main()